File: SR_SVM.R
#Voice/Person Recognition: Support Vector Machines
#Author: Max Weinhold

#In order to create a SVR model with R you will need the package e1071
if(!require("e1071")) install.packages("e1071")
library(e1071)

library(tidyverse)
library(sandwich)
library(caret)

#Clean up memory
rm(list=ls())

#Load Data
setwd("D:/ComicandSonsProductions/GameJam1/SpeakerRecognition/TestData")
VoiceData = read.csv(file = "Test.csv",sep=",")

VoiceData[1]=NULL
names(VoiceData)[1]="Voice"
class(VoiceData$Voice)
VoiceData$Voice = as.factor(VoiceData$Voice)
nrow(VoiceData)

# Split data to reduce duration of computation
training.samples <- VoiceData$Voice %>%
  createDataPartition(p = 0.8, list = FALSE)
train.data  <- VoiceData[training.samples, ]
test.data <- VoiceData[-training.samples, ]

model <- svm(Voice ~., data =  train.data, type = "C-classification", kernel = "polynomial", epsilon = 0.1)

test_predict <- as.data.frame(predict(model, newdata = test.data, type='response'))
train_predict <- as.data.frame(predict(model, data = train.data, type='response'))

EvaluationData<-data.frame(test.data$Voice,test_predict)
names(EvaluationData)<-c("Observation","Prediction")
EvaluationData$Accuracy = EvaluationData$Prediction == EvaluationData$Observation
sum(EvaluationData$Prediction == EvaluationData$Observation)/nrow(EvaluationData)*100

EvaluationData<-data.frame(train.data$Voice,train_predict)
names(EvaluationData)<-c("Observation","Prediction")
EvaluationData$Accuracy = EvaluationData$Prediction == EvaluationData$Observation
sum(EvaluationData$Prediction == EvaluationData$Observation)/nrow(EvaluationData)*100

File: SR_SVM.py
# -*- coding: utf-8 -*-
"""
Created on Sun Sep 25 23:16:01 2022

@author: MaxWe
"""
#About SVM
#Effective in high dimensional spaces.
#Still effective in cases where number of dimensions is greater than the number of samples.
#Uses a subset of training points in the decision function (called support vectors), so it is also memory efficient.
#Versatile: different Kernel functions can be specified for the decision function. Common kernels are provided, but it is also

import matplotlib.pyplot as plt
import numpy as np

#Bibliothek um CSV Dateien zu lesen und zu speichen
import pandas as pd
import csv

#Import SVM Library
#from sklearn import svm
#https://holypython.com/svm/support-vector-machine-step-by-step/

import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split as tts
from sklearn import svm
from sklearn import metrics

#Ließ die CSV Datei, die alle Namen der SoundFiles enthält
with open("D:\ComicandSonsProductions\GameJam1\SpeakerRecognition\TestData\\Frequencies60.csv", newline='') as f:
    reader = csv.reader(f)
    data = list(reader)

voice_data = pd.DataFrame(data)
voice_column = voice_data.iloc[:, 1]
sound_columns = voice_data.iloc[: , 2:]

###Splitting train/test data
X_tr, X_tst, y_tr, y_tst = tts(sound_columns , voice_column , test_size=25/100,random_state=109)

###Creating Support Vector Machine Model
clf = svm.SVC(kernel='rbf', shrinking=True)

###Training the Model
clf.fit(X_tr, y_tr)

###Making Predictions
y_pr = clf.predict(X_tst)
print(y_pr)

###Evaluating Prediction Accuracy
print("Accuracy:",metrics.accuracy_score(y_tst, y_pr))
